### PR TITLE
Ajv 8

### PR DESCRIPTION
**Pull request by Eomm**
This is just a draft PR for ajv8 support - it is a semver major

The interface exposed to fastify is unchanged, but under the hood ajv8 is running 👍🏽 

TODO:

- [x] optimize default configuration
- [x] add tests for most used ajv plugins
- [x] update docs

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -28,12 +28,13 @@
   },
   "homepage": "https://github.com/fastify/ajv-compiler#readme",
   "devDependencies": {
-    "ajv-errors": "^1.0.1",
+    "ajv-errors": "^3.0.0",
+    "ajv-formats": "^2.0.2",
     "ajv-merge-patch": "^4.1.0",
```
Comment (wa-Nadoo) [inline package.json:33]: Seems like 'ajv-merge-patch' is unused

**Comment by mcollina** [comment]
I'm a bit lost on the standalone usage to end users.
The main "client" of this is Fastify itself for its configs.

**Review by mcollina — APPROVED** [review]
lgtm

File: standalone.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  if (options.readMode === true && !options.restoreFunction) {
+    throw new Error('You must provide a restoreFunction options when readMode ON')
+  }
+
+  if (options.readMode !== true && !options.storeFunction) {
+    throw new Error('You must provide a storeFunction options when readMode OFF')
+  }
+
+  if (options.readMode === true) {
+    /*************/
```
Comment (zekth) [inline standalone.js:16]:
```suggestion
    
```

File: standalone.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    // READ MODE: it behalf only in the restore function provided by the user
+    return function wrapper () {
+      return function (opts) {
+        return options.restoreFunction(opts)
+      }
+    }
+  }
+
+  /*************/
```
Comment (zekth) [inline standalone.js:25]:
```suggestion

```

**Review by zekth — CHANGES_REQUESTED** [review]
Minor comment otherwise LGTM.

Thinking out loud: the write/read mode could be exported as a feature of the `fastify-cli` like some test runners have for snapshots. We would be have to do `fastify build --write-schemas` or something. That would be really convinient.

**Comment by Eomm** [comment]
> Thinking out loud: the write/read mode could be exported as a feature of the fastify-cli 

This would be good indeed, but the user needs to change the code to give us the control of the `fastify` instance in some way

**Comment by zekth** [comment]
> > Thinking out loud: the write/read mode could be exported as a feature of the fastify-cli
> 
> This would be good indeed, but the user needs to change the code to give us the control of the `fastify` instance in some way

Maybe that's a feature we could put in the scope of v4. What's your thought on it @mcollina ?

**Review by zekth — APPROVED** [review]

**Comment by mcollina** [comment]
> > > Thinking out loud: the write/read mode could be exported as a feature of the fastify-cli
> 
> > 
> 
> > This would be good indeed, but the user needs to change the code to give us the control of the `fastify` instance in some way
> 
> 
> 
> Maybe that's a feature we could put in the scope of v4. What's your thought on it @mcollina ?

let's go for it! Could you open a fresh issue on fastify-cli?^

**Comment by Eomm** [comment]
I have added tests/examples for

- [x] ajv-formats
- [x] ajv-errors
- [x] ajv-i18n

I think we are ready to publish the v2 of this module and add some docs for Fastify v3

**Review by zekth — APPROVED** [review]

**Review by jsumners — APPROVED** [review]
Looks good to me.

**Pull request merged**